File: src/scene_project_finder.cpp
#include "scene_project_finder.h"
#include "filefinder.h"
#include "sprite.h"
#include "font.h"
#include "input.h"
#include "bitmap.h"
#include "main_data.h"
#include "scene_title.h"
#include "player.h"

#include <cstdlib>

#include <sstream>
#include <algorithm>

#ifdef _WIN32
#  include "windows.h"
#endif


namespace {
std::string get_home_path() {
	return
			std::getenv("HOME")? std::getenv("HOME"):
			std::getenv("HOMEDRIVE") and std::getenv("HOMEPATH")
			? std::string(std::getenv("HOMEDRIVE")).append(std::getenv("HOMEPATH")):
			std::string() // empty string that means home path not found
			;
}

enum { COLUMN_MAX = 25, ROW_MAX = 18, };
}

struct Scene_ProjectFinder::Entry {
	Entry(std::string const& n, std::string const& p)
			: name(n), path(FileFinder().fullpath(p))
			, cursor(0), offset(0)
	{
		assert(not is_project_);
		assert(not path.empty());
	}

	std::string fullpath() const {
		return parent
				? FileFinder().MakePath(parent->fullpath(), path)
				: FileFinder().fullpath(path);
	}

	boost::optional<Entry&> parent;
	std::string const name;
	std::string const path;

	unsigned cursor, offset;

	Sprite& sprite() {
		if(not sprite_.GetBitmap()) {
			children_ = create_children();
			sprite_.SetBitmap(children_.empty()
							  ? Bitmap::Create(1, 1) : create_bitmap(children_));
			sprite_.visible = false;
		}
		return sprite_;
	}

	EntryList const& children() {
		// assert(not children_.empty());
		sprite();
		return children_;
	}

	static bool no_children(EntryRef const& e) {
		assert(e);
		return e->children_.empty();
	}

	bool is_project() {
		FileFinder_& f = FileFinder();
		return *(is_project_ = is_project_? *is_project_ :
				 f.IsRPG2kProject(f.GetDirectoryMembers(
					 fullpath(), FileFinder_::DIRECTORIES).members));
	}

  private:
	EntryList children_;
	Sprite sprite_;
	boost::optional<bool> is_project_;

	EntryList create_children() {
		typedef FileFinder_::Directory Directory;
		FileFinder_& f = FileFinder();

		EntryList ret;
		Directory const dir = f.GetDirectoryMembers(fullpath(), FileFinder_::DIRECTORIES);

		for(FileFinder_::string_map::const_iterator i = dir.members.begin(); i != dir.members.end(); ++i) {
			Directory const d = f.GetDirectoryMembers(
				f.MakePath(dir.base, i->second), FileFinder_::DIRECTORIES);
			if(d.members.empty()) { continue; }

			ret.push_back(EASYRPG_SHARED_PTR<Entry>(new Entry(
				*this, i->second,
				f.IsRPG2kProject(f.GetDirectoryMembers(
					f.MakePath(dir.base, i->second), FileFinder_::FILES).members))));
		}

		return ret;
	}

	Entry(Entry& p, std::string const& pth, bool const is_proj)
			: parent(p), name(pth + (is_proj? "" : "/")), path(pth)
			, cursor(0), offset(0), is_project_(is_proj)
	{
		assert(is_project_);
		assert(not path.empty());
	}
};

BitmapRef Scene_ProjectFinder::create_bitmap(EntryList const& list) {
	assert(not list.empty());

	size_t const
			font_size_ = Font::Shinonome()->pixel_size(),
			width = font_size_ * COLUMN_MAX;
	BitmapRef const ret = Bitmap::Create(width, font_size_ * list.size());
	ret->font = Font::Shinonome();

	for(EntryList::const_iterator i = list.begin(); i < list.end(); ++i) {
		Font::default_color =
				(*i)->is_project()? Color(255, 0, 0, 255) : Color(0, 0, 0, 255);

		int const y = font_size_ * (i - list.begin());
		ret->draw_text(0, y, (*i)->name);

		// fill with .... if directory name is too long
		if(ret->text_size((*i)->name).width > int(width)) {
			ret->fill(Rect(width - font_size_ * 2, y, font_size_ * 2, font_size_), Color(0, 0, 0, 0));
			ret->draw_text(width - font_size_ * 2, y, "....");
		}
	}

	return ret;
}

size_t Scene_ProjectFinder::current_children_count() const {
	return current_entry_? current_entry_->children().size() : root_.size();
}

unsigned& Scene_ProjectFinder::current_index() {
	return current_entry_? current_entry_->cursor : root_index_;
}

unsigned& Scene_ProjectFinder::current_offset() {
	return current_entry_? current_entry_->offset: root_offset_;
}

Sprite& Scene_ProjectFinder::current_sprite() {
	return current_entry_? current_entry_->sprite() : *root_sprite_;
}

void Scene_ProjectFinder::set_index(int const idx) {
	unsigned& dst_idx = current_index();
	unsigned& offset = current_offset();

	dst_idx = (idx + current_children_count()) % current_children_count();
	offset = std::max(0, std::max<int>(
		int(dst_idx) - int(ROW_MAX - 1), std::min(offset, dst_idx)));

	assert(dst_idx < current_children_count());

	cursor_->SetY(font_size_ * (dst_idx - offset + 1) - 1);
	current_sprite().SetSrcRect(Rect(
		0, font_size_ * offset, font_size_ * COLUMN_MAX, font_size_ * ROW_MAX));
}

void Scene_ProjectFinder::select_entry() {
	assert(current_index() < current_children_count());

	current_sprite().visible = false;
	current_entry_ = *(current_entry_? current_entry_->children() : root_)[current_index()];

	if(current_entry_->is_project()) {
		Main_Data::project_path = current_entry_->fullpath();
		Scene::Push(EASYRPG_MAKE_SHARED<Scene_Title>());
		to_parent();
	} else if(current_entry_->children().empty()) {
		to_parent(); // return to parent
	} else {
		current_sprite().SetX(font_size_);
		current_sprite().SetY(font_size_);
		current_sprite().visible = true;
		set_index(current_index());
	}
}

void Scene_ProjectFinder::to_parent() {
	if(current_entry_) {
		current_entry_->sprite().visible = false;
		current_entry_ = current_entry_->parent;

		current_sprite().visible = true;
		set_index(current_index());
	} else {
		Player().exit_flag = true;
	}

}

void Scene_ProjectFinder::Start() {
	font_ = Font::Shinonome();
	font_size_ = font_->pixel_size();

	root_sprite_.reset(new Sprite());
	background_.reset(new Sprite());
	cursor_.reset(new Sprite());

	background_->SetBitmap(Bitmap::Create(SCREEN_TARGET_WIDTH, SCREEN_TARGET_HEIGHT, Color(255, 255, 255, 255)));
	background_->SetZ(-1000);

	root_sprite_->SetBitmap(create_bitmap(root_));
	root_sprite_->SetX(font_size_);
	root_sprite_->SetY(font_size_);
	root_sprite_->SetSrcRect(Rect(0, 0, font_size_ * COLUMN_MAX, font_size_ * ROW_MAX));

	BitmapRef const cursor_bmp = Bitmap::Create(font_size_ * COLUMN_MAX + 4, font_size_ + 2);
	Color const blue(0, 0, 255, 255);
	// top
	cursor_bmp->fill(Rect(0, 0, cursor_bmp->width(), 1), blue);
	// left
	cursor_bmp->fill(Rect(0, 0, 1, cursor_bmp->height()), blue);
	// right
	cursor_bmp->fill(Rect(cursor_bmp->width() - 1, 0, 1, cursor_bmp->height()), blue);
	// bottom
	cursor_bmp->fill(Rect(0, cursor_bmp->height() - 1, cursor_bmp->width(), 1), blue);

	cursor_->SetBitmap(cursor_bmp);
	cursor_->SetX(font_size_ - 2);
	cursor_->SetY(font_size_ - 1);
}

Scene_ProjectFinder::Scene_ProjectFinder()
		: Scene("ProjectFinder")
		, root_index_(0), root_offset_(0)
{
	std::ostringstream current_path;
	current_path << ". (" << FileFinder().fullpath(".") << ")";
	root_.push_back(EASYRPG_MAKE_SHARED<Entry>(current_path.str(), "."));

	std::string const home = get_home_path();
	if(not home.empty()) {
		std::ostringstream oss;
		oss << "HOME (" << home << ")";
		root_.push_back(EASYRPG_MAKE_SHARED<Entry>(oss.str(), home));
	}

#ifdef _WIN32
	do {
		wchar_t drives[MAX_PATH + 1] = {0};
		if(GetLogicalDriveStringsW(MAX_PATH, drives) == 0) {
			Output::Warning("cannot get drives");
			break;
		}

		wchar_t const* ptr = drives;
		while(not std::wstring(ptr).empty()) {
			std::wstring const d = ptr;
			ptr += d.size() + 1;

			wchar_t buf[MAX_PATH + 1] = {0};
			GetVolumeInformationW(d.c_str(), buf, MAX_PATH, NULL, NULL, NULL, NULL, 0);

			std::string const drive = Utils::FromWideString(d);
			std::string const vol = Utils::FromWideString(buf);

			if(vol.empty()) {
				root_.push_back(EASYRPG_MAKE_SHARED<Entry>(drive, drive));
			} else {
				std::ostringstream oss;
				oss << drive << " (" << vol << ")";
				root_.push_back(EASYRPG_MAKE_SHARED<Entry>(drive, oss.str()));
			}
		}
	} while(false);
#else
	root_.push_back(EASYRPG_MAKE_SHARED<Entry>("root (/)", "/"));
#endif

	root_.resize(root_.end() - std::remove_if(root_.begin(), root_.end(), Entry::no_children));
}

void Scene_ProjectFinder::Update() {
	if(Input().IsTriggered(Input_::DECISION) or
	   Input().IsTriggered(Input_::RIGHT)) { select_entry(); }
	else if(Input().IsRepeated(Input_::UP)) { set_index(int(current_index()) - 1); }
	else if(Input().IsRepeated(Input_::DOWN)) { set_index(int(current_index()) + 1); }
	else if(Input().IsTriggered(Input_::CANCEL) or
			Input().IsTriggered(Input_::LEFT)) { to_parent(); }
}

Scene_ProjectFinder::~Scene_ProjectFinder() {}

File: mrblib/scene_project_finder.rb
class Scene_ProjectFinder < Scene
  def initialize
    super "Project Finder"
  end

  def register_project_base_path(ini, section)
    app = Registry.read_string_value(Registry::HKEY_LOCAL_MACHINE,
                                     "Software\\Microsoft\\Windows\\CurrentVersion\\App Paths\\#{ini}.exe",
                                     'Path')
    return unless app

    assert FileFinder.exists? app

    base_path = Registry.read_string_value(FileFinder.make_path(app, ini + ".ini"),
                                           section, "ProjectBasePath")
    @root.push Entry.new('Project Base Path %s (%s)' % [ini, base_path], base_path) if
      not base_path.nil? and FileFinder.exists?(base_path)
  end

  COLUMN_MAX = 25
  ROW_MAX = 18

  class Entry
    def initialize(n, p, is_project = nil)
      @path, @is_project = p, is_project
      @cursor, @offset = 0, 0

      if is_project.nil?
        @parent, @name = nil, n
        assert ! project?
      else
        @parent, @name = n, p + (is_project ? '' : '/')
      end
      assert ! fullpath.empty?
    end

    def fullpath
      @parent.nil? ? FileFinder.fullpath(@path) : FileFinder.make_path(@parent.fullpath, @path)
    end

    attr_accessor :cursor, :offset
    attr_reader :name, :path, :parent

    def sprite
      unless @sprite
        @sprite = Sprite.new
        @children = create_children
        @sprite.bitmap = @children.empty? ? nil : Scene_ProjectFinder.create_bitmap(@children)
        @sprite.visible = false
      end
      @sprite
    end

    def children
      sprite
      return @children
    end

    def project?
      @is_project = FileFinder.rpg2k_project? fullpath if @is_project.nil?
      @is_project
    end

    def create_children
      ret = []
      dir_base = fullpath
      FileFinder.directory_members(dir_base, FileFinder::DIRECTORIES).each do |k,v|
        ret.push Entry.new self, v, FileFinder.rpg2k_project?(FileFinder.make_path(dir_base, v)) unless
          FileFinder.directory_members(FileFinder.make_path(dir_base, v), FileFinder::DIRECTORIES).empty?
      end
      ret
    end
  end

  def update
    if Input.trigger?(Input::DECISION) or Input.trigger?(Input::RIGHT); select_entry
    elsif Input.repeat?(Input::UP); self.index -= 1
    elsif Input.repeat?(Input::DOWN); self.index += 1
    elsif Input.trigger?(Input::CANCEL) or Input.trigger?(Input::LEFT); to_parent
    end
  end

  def index; @current_entry.nil? ? @root_index : @current_entry.cursor; end
  def index=(idx)
    idx = (idx + children_count) % children_count
    @current_entry.nil? ? @root_index = idx : @current_entry.cursor = idx
    self.offset = [0, self.index - ROW_MAX + 1, [self.offset, self.index].min].max

    assert self.index < self.children_count

    @cursor.y = @font_size * (self.index - self.offset + 1) - 1
    self.sprite.src_rect = Rect.new(0, @font_size * self.offset,
                                    @font_size * COLUMN_MAX, @font_size * ROW_MAX)
  end

  def children_count; @current_entry.nil? ? @root.length : @current_entry.children.length end

  def offset; @current_entry.nil? ? @root_offset : @current_entry.offset end
  def offset=(v) @current_entry.nil? ? @root_offset = v : @current_entry.offset = v end

  def sprite; @current_entry.nil? ? @root_sprite : @current_entry.sprite end

  def self.create_bitmap(list)
    assert !list.empty?

    font_size = 12 # Font.shinonome.pixel_size
    ret = Bitmap.new font_size * COLUMN_MAX, font_size * list.length

    list.each_with_index do |v,i|
      Font.default_color = Color.new(v.project? ? 255 : 0, 0, 0, 255)
      ret.draw_text 0, font_size * i, v.name
    end

    ret
  end

  def select_entry
    assert index < children_count

    sprite.visible = false
    @current_entry = (@current_entry.nil? ? @root : @current_entry.children)[index]

    if @current_entry.project?
      FileFinder.project_path = @current_entry.fullpath
      Player.push Scene_Title.new
      to_parent
    elsif @current_entry.children.empty?
      to_parent # return to parent
    else
      sprite.x, sprite.y = @font_size, @font_size
      sprite.visible = true
      self.index = index
    end
  end

  def to_parent
    if @current_entry.nil?
      Player.exit_flag = true
    else
      @current_entry.sprite.visible = false
      @current_entry = @current_entry.parent
      sprite.visible = true
      self.index = index
    end
  end

  def start
    @current_entry = nil

    @root = []
    @root_index, @root_offset = 0, 0
    @root_sprite, @background, @cursor = nil, nil, nil

    @font_size = 0

    @root.push Entry.new(". (#{FileFinder.fullpath('.')})", '.')

    home = Utils.home_path
    @root.push Entry.new('HOME (%s)' % home, home) if not home.nil?

    Utils.drives.each do |k,v|
      @root.push Entry.new(k.to_s + v.empty? ? '' : ' (#{v})', k.to_s)
    end

    register_project_base_path "RPG2000T", "RPG2000"
    register_project_base_path "RPG2000" , "RPG2000"

    @root = @root.reject { |v| v.children.empty? }

    @font_size = 12 # @font.pixel_size

    @root_sprite, @background, @cursor = Sprite.new, Sprite.new, Sprite.new

    @background.bitmap = Bitmap.new SCREEN_TARGET_WIDTH, SCREEN_TARGET_HEIGHT
    @background.bitmap.fill_rect @background.bitmap.rect, Color.new(255, 255, 255, 255)
    @background.z = -1000

    @root_sprite.bitmap = Scene_ProjectFinder.create_bitmap @root
    @root_sprite.x, @root_sprite.y = @font_size, @font_size
    @root_sprite.src_rect = Rect.new 0, 0, @font_size * COLUMN_MAX, @font_size * ROW_MAX

    cursor_bmp = Bitmap.new @font_size * COLUMN_MAX + 4, @font_size + 2
    blue = Color.new 0, 0, 255, 255
    # top
    cursor_bmp.fill_rect Rect.new(0, 0, cursor_bmp.width, 1), blue
    # left
    cursor_bmp.fill_rect Rect.new(0, 0, 1, cursor_bmp.height), blue
    # right
    cursor_bmp.fill_rect Rect.new(cursor_bmp.width - 1, 0, 1, cursor_bmp.height), blue
    # bottom
    cursor_bmp.fill_rect Rect.new(0, cursor_bmp.height - 1, cursor_bmp.width, 1), blue

    @cursor.bitmap = cursor_bmp
    @cursor.x = @font_size - 2
    @cursor.y = @font_size - 1
  end
end
